subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalInputsIntegrationTest.groovy

```diff
@@ -73,4 +105,45 @@ class IncrementalInputsIntegrationTest extends AbstractIncrementalTasksIntegrati
         executesWithRebuildContext()
     }
 
+    def "cannot query non-incremental file input parameters"() {
+        given:
+        previousExecution()
+
+        when:
+        file("inputs/new-input-file.txt") << "new file"
+        buildFile << """
+            tasks.withType(IncrementalTask).configureEach {
+                queryChangesFor.add("nonIncrementalInput")
+            }
+        """
+        then:
+        fails("incremental")
+        failure.assertHasCause("Cannot query incremental changes: No property found for value ${file("nonIncrementalInput").absolutePath}. Incremental properties: anotherIncrementalInput, inputDir.")
+    }
+
+    def "changes to non-incremental input parameters cause a rebuild"() {
+        given:
+        file("nonIncrementalInput").makeOlder()
+        previousExecution()
+
+        when:
+        file("inputs/new-input-file.txt") << "new file"
+        file("nonIncrementalInput").text = 'changed'
+        then:        
+        executesWithRebuildContext("ext.added += ['new-input-file.txt']")
```

Comment:
This took a bit for me to decipher. For one, `ext.added` doesn't really convey that these are the expectations we set. This would be much more readable for me:

```groovy
executesNonIncrementally() // when there are no additional files to expect
executesNonIncrementallyWithAdditionalInputs('new-input-file.txt') // when there are no additional files to expect
```